variables.tf

```diff
@@ -40,6 +47,7 @@ variable "allowed_ports" {
 
 variable "subnet" {
   description = "VPC Subnet ID the instance is launched in"
+  default     = ""
```

Comment:
Since `subnet` is a required field, I don't think we should add a default.

Reply:
Aha, I see you made it dynamic. 

https://github.com/cloudposse/terraform-aws-ec2-instance/pull/28/files#diff-7a370d8342e7203b805911c92454f0f4R20

Heads up, there's a *huge* caveat with using `data` providers. They introduce race conditions. Using a `data` provider means that you cannot just run `terraform apply` without targeting a specific module during the apply. This is because `data` providers are executed as part of the `plan` phase and cannot depend on resources created as part of the `apply` phase.

Reply:
So remove the default value, or leave it with data provider?

Reply:
I'm inclined to say leave it out as it will only lead to "new issues" getting filed =)

@aknysh @goruha  do you have any stronger opinions?

Reply:
I think we should not use the data providers (unless really necessary, e.g. lookup other VPC attributes by the provided VPC ID) in the lower-level modules as it could create race conditions when used in top-level modules.
Also, it's not obvious to the users of the modules what's going on.
It's better to do all the lookups (if needed) in top-level modules and just provide the values to the lower-level modules.
More explicit, less magic.

Reply:
I do not like `dynamic subnet`. Let's make `subnet` field required and remove all `dynamic subnet` staff